### PR TITLE
fix(controller): fix panic when last promotion has no freight collection

### LINKED ISSUE #2361: panic when promoting freight
# Checklist

* [x] I've searched the issue queue to verify this is not a duplicate bug report.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `kargo version`.
* [x] I've pasted logs, if applicable.

# Description

New to Kargo and deployed a small setup but encountered a nil error. Not sure what the issue is, might be a misconfiguration on my part but would not expect silent crash. Created a PR https://github.com/akuity/kargo/pull/2360 to give more and raising this to track the issue.

# Screenshots

Screenshot 2024-07-29 at 14 43 25

# Steps to Reproduce

- Install fresh Kargo on Docker Desktop using [Hacking on Kargo](https://kargo.akuity.io/contributor-guide/hacking-on-kargo) instructions
- Create `Warehouse`
- Create `Stage`

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: dev
  namespace: kargo-demo
spec:
  promotionMechanisms:
    argoCDAppUpdates:
      - appName: go-hello-world
    gitRepoUpdates:
      - kustomize:
          images:
            - image: ghcr.io/loafoe/go-hello-world
              path: stages/dev
              useDigest: false
        repoURL: https://github.com/loafoe/kargo-demo.git
        writeBranch: main
  requestedFreight:
    - origin:
        kind: Warehouse
        name: kargo-demo
      sources:
        direct: true
  subscriptions: {}
```

# Version

```
Client Version: v0.8.0
Server Version: v0.8.1
```

# Logs

```
time="2024-07-29T10:15:59Z" level=debug msg="found associated Stage" freight=89070859a69cdbd179a6a33e9cb6d2f77e1c5b83 namespace=kargo-demo promotion=dev.01j3yztqpsmxmbm3230grxv8yt.8907085 stage=dev
time="2024-07-29T10:15:59Z" level=error msg="Promotion panic" error="runtime error: invalid memory address or nil pointer dereference" 
```

# Stacktrace (captured through PR)

```
goroutine 289 [running]:
runtime/debug.Stack()
	/opt/homebrew/Cellar/go/1.22.5/libexec/src/runtime/debug/stack.go:24 +0x64
github.com/akuity/kargo/internal/controller/promotions.(*reconciler).Reconcile.func4.1()
	/Users/andy/DEV/Go/kargo/internal/controller/promotions/promotions.go:348 +0x80
panic({0x1ab0080?, 0x34eafe0?})
	/opt/homebrew/Cellar/go/1.22.5/libexec/src/runtime/panic.go:770 +0x124
github.com/akuity/kargo/internal/controller/promotions.(*reconciler).buildTargetFreightCollection(0x4000820010?, {{0x40002fb8f0, 0x28}, {0x0, 0x0}, {{0x40002c72e4, 0x9}, {0x40002c72f0, 0xa}}, {0x0, ...}, ...}, ...)
	/Users/andy/DEV/Go/kargo/internal/controller/promotions/promotions.go:548 +0x1d4
github.com/akuity/kargo/internal/controller/promotions.(*reconciler).promote(_, {_, _}, {{{0x1870b48, 0x9}, {0x40008283c0, 0x18}}, {{0x4000951bc0, 0x26}, {0x0, ...}, ...}, ...}, ...)
	/Users/andy/DEV/Go/kargo/internal/controller/promotions/promotions.go:490 +0x4fc
github.com/akuity/kargo/internal/controller/promotions.(*reconciler).Reconcile.func4(0x40003e5b00, 0x40003e5af0, 0x1aafc00?, {0x228cc20?, 0x4000902f00?}, 0x4000d10d20?, 0x4000a0bc20?, 0x35554aaaa?)
	/Users/andy/DEV/Go/kargo/internal/controller/promotions/promotions.go:356 +0xd4
github.com/akuity/kargo/internal/controller/promotions.(*reconciler).Reconcile(0x4000234a10, {0x228cc20, 0x4000902ab0}, {{{0x4000a0b9a6?, 0x40003e5cc8?}, {0x4000951ad0?, 0x40003e5d08?}}})
	/Users/andy/DEV/Go/kargo/internal/controller/promotions/promotions.go:369 +0x98c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x2294b00?, {0x228cc20?, 0x4000902ab0?}, {{{0x4000a0b9a6?, 0xb?}, {0x4000951ad0?, 0x0?}}})
	/Users/andy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:114 +0x8c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0x40007c7340, {0x228cc58, 0x40000dd180}, {0x1c110a0, 0x4000d10980})
	/Users/andy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:311 +0x2dc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x40007c7340, {0x228cc58, 0x40000dd180})
	/Users/andy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:261 +0x16c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/Users/andy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:222 +0x74
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 128
	/Users/andy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:218 +0x3bc
```

### PR DESCRIPTION
**Pull request by loafoe**
- Fixes crash by adding additional nil checks
- Resolves #2361

### CODE REVIEW

File: internal/controller/promotions/promotions.go [hunk 1]
```diff
@@ -544,6 +546,10 @@ func (r *reconciler) buildTargetFreightCollection(
 	lastPromo := stage.Status.LastPromotion
 	if lastPromo != nil {
 		for _, req := range stage.Spec.RequestedFreight {
+			if lastPromo.Status == nil || lastPromo.Status.FreightCollection == nil || lastPromo.Status.FreightCollection.Freight == nil {
```
Comment (hiddeco) [inline internal/controller/promotions/promotions.go:549]:
```suggestion
			if lastPromo.Status == nil ||
				lastPromo.Status.FreightCollection == nil ||
				lastPromo.Status.FreightCollection.Freight == nil {
```

to please the linter.

**Comment by loafoe** [comment]
@hiddeco thanks for reviewing, rebased to make sure DCO passes (signature)

File: internal/controller/promotions/promotions.go [hunk 2]
```diff
@@ -544,6 +546,10 @@ func (r *reconciler) buildTargetFreightCollection(
 	lastPromo := stage.Status.LastPromotion
 	if lastPromo != nil {
 		for _, req := range stage.Spec.RequestedFreight {
+			if lastPromo.Status == nil || lastPromo.Status.FreightCollection == nil {
+				logger.Error(nil, "last promotion has no freight collection")
```
Comment (hiddeco) [inline internal/controller/promotions/promotions.go:550]:
```suggestion
				logger.Debug("last promotion has no freight collection")
```

Comment (hiddeco) [reply internal/controller/promotions/promotions.go:550]: Sorry, this went unnoticed.

Comment (loafoe) [reply internal/controller/promotions/promotions.go]: Fixed

**Pull request merged**